USDC-on-Polygon integration

sisou opened the pull request:
Generate Polygon address(es) when creating a new account or logging in, and add request to sign OpenGSN relay requests for transferring USDC via Nimiq's USDC-transfer-contract. Also add USDC support to sign-swap request(s).

sisou committed:
Return Polygon address from Create request

sisou committed:
Return Polygon address from Import request

sisou committed:
Implement PolygonKey using ethers.js

TODOs:
- Copying ethers.js into dist folder and replacing script path

sisou committed:
Update SignPolygonTransaction request to work with Nimiq's OpenGSN contract

onmax committed:
Renamed openGSN contract to NimiqUSDC contract

sisou committed:
Fix USDC amounts losing decimal precision when converting from units

Reason is that ethers.BigNumber cannot represent decimals, only integers.

sisou committed:
Remove unnecessary CSS rule that is also included in @nimiq/style

sisou committed:
Display fee (in USDC) for USDC transactions

sisou committed:
Add TypedRequestData from @opengsn/common as an UMD module

Converted with browserify.
Use as `new OpenGSN.TypedRequestData()`

sisou committed:
Add types for OpenGSN module

sisou committed:
Sign OpenGSN RelayRequest typed data for USDC transactions

sisou committed:
Accept an OpenGSN RelayRequest for signing

Instead of a Polygon transaction.
Misses parsing of RelayRequest, is currently trusting the input.

sisou committed:
Update USDC Transfer contract ABI

sisou committed:
Add missing dependency for client

sisou committed:
Add DerivePolygonAddress request

sisou committed:
Add SignPolygonTransaction request

sisou committed:
Update create request with new onboarding flow

Remove identicon selection, instead focus more on explaining Login Files.

sisou committed:
Fix Gitlab CI config for branch previews

sisou committed:
Do not bundle the OpenGSN lib in polygon-bundle

sisou committed:
Add USDC support to sign-swap request

sisou committed:
Make SignSwap slider layout generic over the swap pair

sisou committed:
Fix USDC parsing for swaps

sisou committed:
Fix adding approval signature to openWithApproval method data

sisou committed:
Fix contract address that gets approved for swaps

sisou committed:
Fix bug when swapping to USDC

sisou committed:
Fix linter errors

sisou committed:
Merge branch 'soeren/usdc/swap' into usdc

sisou committed:
Update wording in derive-polygon-address request

Also migrate transifex config to new client.

sisou committed:
Update USDC contract ABI

[skip ci]

onmax committed:
updated swap slider layout

edit: signed commit

sisou committed:
Small styling & formatting adjustments and fix CI

sisou committed:
Update USDC transfer contract to v3

danimoh committed:
ServiceWorker: preserve testnet branch selector cookie

sisou committed:
Set mainnet transfer contract address

sisou committed:
Adjust SignSwap slider layout

sisou committed:
Remove chainTokenFee from Transfer args interfaces

Forgotten as part of 25cdb4a9e8f2cfaf45549f08b75ee70a90d45d0d.

sisou committed:
Remove unnecessary fee property from USDC swap redeem details

sisou committed:
Allow signing USDC HTLC refund transactions

sisou committed:
Update USDC HTLC contract to v3

sisou committed:
Move common Polygon parsers into a common PolygonRequestParserMixin

sisou committed:
Add more Polygon lib files to the polygon-bundle

sisou committed:
Parse OpenGSN request parts

sisou committed:
Set mainnet HTLC contract address

[skip ci]

sisou committed:
Update translations

mraveux commented on client/src/PublicRequest.ts:
```diff
@@ -313,13 +361,18 @@ export type SignSwapRequestStandard = SignSwapRequestCommon & {
     },
+    polygonAddresses: Array<{
+        address: string,
+        balance: number, // cents
```
Is it really in cents? If yes, then it would be the first time we're not using the smallest unit, right?
And at the same time it's a bit counter intuitive to use cents for polygon (which use Matic and Wei).

Just pointing it out, not asking for changes.

sisou replied:
Oh, no, you are right, it's not in Cents, it's in USDC smallest unit. What should we call that?

mraveux replied:
It apparently doesn't have a name, but Polygon and ETH smallest unit is Wei, so maybe we can call it usdc-wei or smth the like.

mraveux committed:
DerivePolygonAddress demo: minor improvements to the demo page

mraveux commented on src/lib/polygon/PolygonUtils.js:
```diff
@@ -0,0 +1,41 @@
+     * @param {number} coins USDC coins in decimal
+     * @returns {number} Number of Cents
+     */
+    static coinsToCents(coins) {
```
I guess this comment follow the previous one about cents. We can keep it like that even if it's not exactly what's happenning (since it's not cents but 1/10000 of a cent iirc, and "cents" literally means "a hundredth") or change it for consistency to something like `weiToCoins`, `usdcweiToCoins`, `millionthToCoins` or anything else more correct than `cents`.

danimoh commented on client/src/PublicRequest.ts:
```diff
@@ -1,6 +1,13 @@
 import * as Nimiq from '@nimiq/core-web';
+import { ForwardRequest } from '@opengsn/common/dist/EIP712/ForwardRequest';
+import { RelayData } from '@opengsn/common/dist/EIP712/RelayData';
```
Should have more specific names, e.g. `OpenGsnForwardRequest` and `OpenGsnRelayData`.

danimoh commented on client/src/PublicRequest.ts:
```diff
@@ -25,6 +32,10 @@ export type SingleKeyResult = {
     wordsExported: boolean;
     bitcoinXPub?: string;
+    polygonAddresses?: Array<{
+        address: string, // Could be a Uint8Array by decoding the address as HEX?
```
Both are fine. Just make a choice and remove the comment.

danimoh commented on client/src/PublicRequest.ts:
```diff
@@ -268,6 +310,12 @@ export type SignSwapRequestCommon = SimpleRequest & {
+        {type: 'USDC'}
+        & Omit<PolygonTransactionInfo, 'approval'>
+        & {
+            amount: number,
```
`PolygonTransactionInfo` already contains optional `amount?: number`. Should the optional `amount` be omitted first or does it make no difference?

danimoh commented on client/src/PublicRequest.ts:
```diff
@@ -313,13 +361,18 @@ export type SignSwapRequestStandard = SignSwapRequestCommon & {
     bitcoinAccount: {
         balance: number, // Sats
     },
+    polygonAddresses: Array<{
```
As the balance does not refer to Polygon and the swapped currency is not Polygon, this should be called `usdcAddresses` or `polygonUsdcAddresses` or `usdcPolygonAddresses` or the balance be called `usdcBalance`.

sisou replied:
I chose to go with changing `balance` to `usdcBalance`, because the address is a Polygon address, and USDC is a token on that Polygon address.

danimoh commented on client/src/PublicRequest.ts:
```diff
@@ -353,7 +409,12 @@ export type SignSwapTransactionsRequest = {
         transactionHash: string,
         outputIndex: number;
-    } | {
+    }  | {
```
Spacing is off.

danimoh commented on client/src/PublicRequest.ts:
```diff
@@ -353,7 +409,12 @@ export type SignSwapTransactionsRequest = {
+        hash: string,
+        timeout: number,
+        htlcId: string,
+    }| {
```
Spacing is off.

danimoh commented on client/tsconfig.json:
```diff
@@ -13,6 +13,7 @@
     "baseUrl": ".",
     "declaration": true,
     "declarationDir": "./dist/src/",
+    "skipLibCheck": true,
```
Consider adding a comment why this is being skipped now, together with a TODO to remove this flag again, once we update Typescript.

danimoh commented on client/src/PublicRequest.ts:
```diff
@@ -1,6 +1,13 @@
 import * as Nimiq from '@nimiq/core-web';
+import { ForwardRequest } from '@opengsn/common/dist/EIP712/ForwardRequest';
+import { RelayData } from '@opengsn/common/dist/EIP712/RelayData';
```
If types suffice, consider `import type` only, if supported by the client's typescript version, or `type ... = import(...)`.

sisou replied:
`import type` is not supported by the current TS version.

sisou replied:
The imports are TS `interface`s, so they are removed by the TS compiler automatically and not included in the output bundle (I checked).

danimoh commented on demos/DerivePolygonAddress.html:
```diff
@@ -0,0 +1,61 @@
+            appName: 'Keyguard Demo',
+            keyId: id,
+            keyLabel: 'My Demo Account',
+            polygonAccountPath: `m/44'/699'/0'`,
```
We're now using the ethereum path by default.

sisou replied:
Only in mainnet. Testnet uses the Polygon SLIP-44 path.

danimoh commented on demos/DerivePolygonAddress.html:
```diff
@@ -0,0 +1,61 @@
+        }
+
+        keyguard.close();
+        await KeyStore.instance.remove(request.keyId);
```
Can be put into a `finally` block.

sisou replied:
But doesn't have to. There's no functional difference.

danimoh commented on package.json:
```diff
@@ -7,8 +7,9 @@
     "serve": "yarn pre && node serve.js",
     "build": "yarn pre && tools/build.sh",
     "build:bitcoinjs": "yarn --silent browserify tools/bitcoinjs-parts.js -s BitcoinJS | yarn terser --compress --mangle --source-map --output src/lib/bitcoin/BitcoinJS.js",
+    "build:opengsn": "yarn browserify -r @opengsn/common/dist/EIP712/TypedRequestData -s OpenGSN -o src/lib/polygon/OpenGSN.js",
```
Can bitcoinjs and opengsn share common dependencies?

sisou replied:
Possibly, if you want to dive into that.

danimoh replied:
Opened an issue for it here: https://github.com/nimiq/keyguard/issues/472

danimoh commented on package.json:
```diff
@@ -7,8 +7,9 @@
     "serve": "yarn pre && node serve.js",
     "build": "yarn pre && tools/build.sh",
     "build:bitcoinjs": "yarn --silent browserify tools/bitcoinjs-parts.js -s BitcoinJS | yarn terser --compress --mangle --source-map --output src/lib/bitcoin/BitcoinJS.js",
+    "build:opengsn": "yarn browserify -r @opengsn/common/dist/EIP712/TypedRequestData -s OpenGSN -o src/lib/polygon/OpenGSN.js",
```
Minify oppengsn.

danimoh commented on src/assets/icons/polygon-white.svg:
```diff
@@ -0,0 +1,3 @@
+<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 38.4 33.5">
+  <path d="M29 10.2c-.7-.4-1.6-.4-2.4 0L21 13.5l-3.8 2.1-5.5 3.3c-.7.4-1.6.4-2.4 0L5 16.3a2.4 2.4 0 0 1-1.2-2.1v-5c0-.8.4-1.6 1.2-2.1l4.3-2.5c.7-.4 1.6-.4 2.4 0L16 7.2c.7.4 1.2 1.2 1.2 2.1v3.3l3.8-2.2V7c0-.8-.4-1.6-1.2-2.1l-8-4.7c-.7-.4-1.6-.4-2.4 0L1.2 5C.4 5.4 0 6.2 0 7v9.4c0 .8.4 1.6 1.2 2.1l8.1 4.7c.7.4 1.6.4 2.4 0l5.5-3.2 3.8-2.2 5.5-3.2c.7-.4 1.6-.4 2.4 0l4.3 2.5c.7.4 1.2 1.2 1.2 2.1v5c0 .8 ...
+</svg>
```
This file is unused.

danimoh commented on src/assets/icons/polygon.svg:
```diff
@@ -0,0 +1,3 @@
+<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 38.4 33.5">
+  <path d="M29 10.2c-.7-.4-1.6-.4-2.4 0L21 13.5l-3.8 2.1-5.5 3.3c-.7.4-1.6.4-2.4 0L5 16.3a2.4 2.4 0 0 1-1.2-2.1v-5c0-.8.4-1.6 1.2-2.1l4.3-2.5c.7-.4 1.6-.4 2.4 0L16 7.2c.7.4 1.2 1.2 1.2 2.1v3.3l3.8-2.2V7c0-.8-.4-1.6-1.2-2.1l-8-4.7c-.7-.4-1.6-.4-2.4 0L1.2 5C.4 5.4 0 6.2 0 7v9.4c0 .8.4 1.6 1.2 2.1l8.1 4.7c.7.4 1.6.4 2.4 0l5.5-3.2 3.8-2.2 5.5-3.2c.7-.4 1.6-.4 2.4 0l4.3 2.5c.7.4 1.2 1.2 1.2 2.1v5c0 .8 ...
+</svg>
```
This file is unused.

danimoh commented on src/components/BalanceDistributionBar.js:
```diff
@@ -2,68 +2,67 @@
+            newBalance: this._unitsToFiat(segment.newBalance, leftAsset, leftFiatRate),
+            backgroundClass: leftAsset === 'NIM'
+                ? LoginFileConfig[IqonHash.getBackgroundColorIndex(segment.address)].className
+                : segment.address,
```
```suggestion
                : leftAsset.toLowerCase(),
```

While the `address` is set to the currency, this is definitely confusing code. Better to use `leftAsset`.

Similar for `rightDistributionData` below.

danimoh commented on src/components/BalanceDistributionBar.js:
```diff
@@ -2,68 +2,67 @@
-        const totalBalance = nimDistributionData.reduce((sum, data) => sum + data.newBalance, 0)
-            + btcDistributionData.newBalance;
+        const totalBalance = leftDistributionData.reduce((sum, data) => sum + data.newBalance, 0)
+            + rightDistributionData.reduce((sum, data) => sum + data.newBalance, 0);
```
```suggestion
        const totalBalance = [...leftDistributionData, ...rightDistributionData].reduce((sum, data) => sum + data.newBalance, 0);
```

danimoh commented on src/components/BalanceDistributionBar.js:
```diff
@@ -103,4 +102,20 @@ class BalanceDistributionBar { // eslint-disable-line no-unused-vars
+            case 'EUR': return EuroUtils.centsToCoins(units) * rate;
+            default: throw new Error('Invalid asset for unit to fiat conversion');
+        }
+    }
```
Could be moved to a shared utils file as `SwapFeesTooltip` also essentially does the same for converting cypto to fiat. Also `SignSwap` does similar stuff.

danimoh commented on src/components/PolygonAddressInfo.js:
```diff
@@ -0,0 +1,88 @@
+     * @param {HTMLElement} [$el]
+     * @returns {HTMLElement}
+     */
+    render($el) {
```
```suggestion
    renderTo($el) {
```

Should be called `renderTo`, congruent to the naming in `AddressInfo`.

danimoh commented on src/components/PolygonAddressInfo.js:
```diff
@@ -0,0 +1,88 @@
+            const $avatar = document.createElement('div');
+            $avatar.classList.add('avatar');
+            if (this._label) {
+                $avatar.textContent = this._label.substring(0, 1).toUpperCase();
```
```suggestion
                $avatar.textContent = this._label[0].toUpperCase();
```

sisou replied:
Why?

danimoh commented on src/components/PolygonAddressInfo.js:
```diff
@@ -0,0 +1,88 @@
+            $img.classList.add('logo');
+            $img.src = '../../assets/icons/unknown.svg';
+            $el.appendChild($img);
+        }
```
```suggestion
        } else if (this._logo === 'usdc' || this._logo === 'unknown') {
            const $img = document.createElement('img');
            $img.classList.add('logo');
            $img.src = `../../assets/icons/${this._logo}.svg`;
            $el.appendChild($img);
        }
```
Both cases can be combined.

Additionally, using a `background-image` instead of a separate image element could be considered, especially as the unknown icon already exists in the css via the `unlabelled` class.

sisou replied:
If the image already exists, then why join them? Just change the unknown case to use the `unlabelled` class. Combining these cases will just lead to us having to separate them again when things change.

sisou replied:
I disagree with this. This is not an optimization in my eyes.

danimoh commented on src/components/PolygonAddressInfo.js:
```diff
@@ -0,0 +1,88 @@
+        if (this._label) {
+            $label.textContent = this._label;
+        } else {
+            $label.textContent = I18n.translatePhrase('label-unknown');
```
```suggestion
            $label.textContent = I18n.translatePhrase('label-unknown');
            $label.dataset.i18n = 'label-unknown';
```
Also set i18n dataset to support reactive update on language switch.

danimoh commented on src/components/PolygonAddressInfo.js:
```diff
@@ -0,0 +1,88 @@
+        $dots.classList.add('dots');
+        $prefix.textContent = this._address.substring(0, 4);
+        $suffix.textContent = this._address.substring(this._address.length - 4);
+        $dots.innerHTML = new Array(3).fill(TemplateTags.noVars`
+            <svg viewBox="0 0 3 3" xmlns="http://www.w3.org/2000/svg">
+                <circle cx="1.5" cy="1.5" r="1.5" fill="currentColor"/>
+            </svg>
+        `.trim()).join('');
```
Could also be a single svg with the three circles inside. Would save a few DOM elements, but it's a minor optimization.

sisou replied:
This way we can space them flexibly.

danimoh commented on src/components/SwapFeesTooltip.js:
```diff
@@ -261,6 +282,24 @@ class SwapFeesTooltip { // eslint-disable-line no-unused-vars
         return [$div];
     }
 
+    /**
+     * @param {number} fiatFee
+     * @param {string} fiatCurrency
+     * @returns {[HTMLDivElement]}
+     */
+    _createUsdcLine(fiatFee, fiatCurrency) {
+        const $div = document.createElement('div');
+        $div.classList.add('price-breakdown');
+
+        $div.innerHTML = TemplateTags.hasVars(1)`
+            <label data-i18n="sign-swap-usdc-fees">USDC network fee</label>
+            <div>${NumberFormatting.formatCurrency(fiatFee, fiatCurrency)}</div>
+        `;
+        I18n.translateDom($div);
+
+        return [$div];
+    }
```
These createLine methods all essentially do the same and can be unified.

sisou replied:
I think it's good they are separate, as some of them have additional explainer paragraphs as well. This way we stay flexible.

danimoh commented on src/lib/NumberFormatting.js:
```diff
@@ -12,9 +12,18 @@ class NumberFormatting { // eslint-disable-line no-unused-vars
+            } else {
+                break;
+            }
+        }
```
```suggestion
        while (result.split('.')[1] && result.split('.')[1].length > minDecimals && result[result.length - 1] === '0') {
            result = result.replace(/\.?0$/, '');
        }
```
Simplify the loop and avoid potential dangling decimal separator.

danimoh commented on src/lib/NumberFormatting.js:
```diff
@@ -12,9 +12,18 @@ class NumberFormatting { // eslint-disable-line no-unused-vars
+        let result = parseFloat(value.toFixed(minDecimals)) === value
             ? value.toFixed(minDecimals)
-            : value.toString();
+            : value.toFixed(maxDecimals);
```
As the loop was added, this check is not required for decimal handling anymore and thus, these lines can be reduced to just `let result = value.toFixed(maxDecimals);`. Note that the rounding above is still required though as `toFixed`'s rounding behavior differs from what we'd want, e.g. `(0.15).toFixed(1)` rounds to `0.1` instead of `0.2`.

What was the issue with the previous approach? Did you have cases where `toString` rendered to scientific notation or rendered too many decimals or where the equality check failed due to floating point imprecision?

sisou replied:
I think it was scientific notation.

danimoh commented on src/lib/polygon/PolygonConstants.js:
```diff
@@ -0,0 +1,4 @@
+const PolygonConstants = { // eslint-disable-line no-unused-vars
+    CENTS_PER_COINS: 1e6,
```
It's not cents.

sisou replied:
It doesn't have a name.

sisou replied:
Renamed to `UNITS_PER_COIN`.

danimoh commented on src/lib/polygon/PolygonConstants.js:
```diff
@@ -0,0 +1,4 @@
+const PolygonConstants = { // eslint-disable-line no-unused-vars
+    CENTS_PER_COINS: 1e6,
+    DEFAULT_DERIVATION_PATH: 'm/44\'/966\'/0\'/0/0', // Used in tests
```
Shouldn't we change this to the ethereum path?

sisou replied:
Doesn't make a difference.

sisou replied:
...as it's only used in tests. For actual requests, the path is provided by the external caller.

danimoh commented on src/lib/polygon/PolygonKey.js:
```diff
@@ -0,0 +1,107 @@
+     */
+    set hasPin(hasPin) {
+        this._key.hasPin = hasPin;
+    }
```
I believe all keys with pins were legacy accounts with legacy mnemonics? In that case, these methods could be omitted. Same actually for Bitcoin.

danimoh commented on src/lib/polygon/PolygonRequestParserMixin.js:
```diff
@@ -0,0 +1,117 @@
+         * @returns {string}
+         */
+        parsePolygonAddress(address, name) {
+            if (typeof address !== 'string' || address.substring(0, 2) !== '0x' || address.substring(2).length !== 40) {
```
```suggestion
            if (typeof address !== 'string' || !/^0x[\dA-Fa-f]{40}$/.test(address)) {
```
Simplify check and also test for allowed characters.
Could in fact also use ethers directly to also verify the address checksum.

danimoh commented on src/lib/polygon/PolygonRequestParserMixin.js:
```diff
@@ -0,0 +1,117 @@
+
+            if (!digitsOnlyRegex.test(request.validUntil)) {
+                throw new Errors.InvalidRequestError('Invalid request.validUntil');
+            }
```
```suggestion
            /** @param {unknown} value */
            const isIntegerString = value => typeof value === 'string' && /^\d+$/.test(value);

            if (!isIntegerString(request.gas)) {
                throw new Errors.InvalidRequestError('Invalid request.gas');
            }

            if (!isIntegerString(request.nonce)) {
                throw new Errors.InvalidRequestError('Invalid request.nonce');
            }

            if (!isIntegerString(request.validUntil)) {
                throw new Errors.InvalidRequestError('Invalid request.validUntil');
            }
```

Verify that the checked values are also strings.

danimoh commented on src/lib/polygon/PolygonRequestParserMixin.js:
```diff
@@ -0,0 +1,117 @@
+        }
+
+        /**
+         *
```
Spacing

danimoh commented on src/lib/polygon/PolygonRequestParserMixin.js:
```diff
@@ -0,0 +1,117 @@
+        }
+
+        /**
+         * @param {KeyguardRequest.ForwardRequest} request
```
Should be of type `unknown`.

danimoh commented on src/lib/polygon/PolygonRequestParserMixin.js:
```diff
@@ -0,0 +1,117 @@
+
+        /**
+         *
+         * @param {KeyguardRequest.RelayData} relayData
```
Should be of type `unknown`.

danimoh commented on src/lib/polygon/PolygonRequestParserMixin.js:
```diff
@@ -0,0 +1,117 @@
+            this.parsePolygonAddress(relayData.paymaster, 'relayData.paymaster');
+            this.parsePolygonAddress(relayData.relayWorker, 'relayData.relayWorker');
+
+            const digitsOnlyRegex = /^[0-9]+$/;
```
See above.

danimoh commented on src/lib/polygon/PolygonRequestParserMixin.js:
```diff
@@ -0,0 +1,117 @@
+                throw new Errors.InvalidRequestError('Invalid request.validUntil');
+            }
+
+            return request;
```
`data` is not verified.

sisou replied:
It is by parsing it into the `description` in the actual requestApi class.

danimoh commented on src/lib/polygon/PolygonUtils.js:
```diff
@@ -0,0 +1,41 @@
+     * @returns {boolean}
+     */
+    static isValidPath(path) {
+        if (path.match(/^m(\/[0-9]+'?)*$/) === null) return false;
```
```suggestion
        if (!/^m(\/[0-9]+'?)*$/.test(path)) return false;
```

sisou replied:
Why?

danimoh commented on src/request/create/index.html:
```diff
@@ -46,19 +46,18 @@
     <script defer bundle-bitcoin src="../../lib/bitcoin/BitcoinKey.js"></script>
     <script defer bundle-bitcoin src="../../lib/bitcoin/BitcoinRequestParserMixin.js"></script>
 
+    <script defer bundle-polygon src="../../../node_modules/ethers/dist/ethers.umd.js"></script>
```
This should be loaded only into a sandboxed iframe.

We might then also consider minifying it.

danimoh replied:
Opened an issue for it here: https://github.com/nimiq/keyguard/issues/471

danimoh commented on src/request/create/Create.js:
```diff
@@ -66,47 +59,47 @@ class Create {
+            let prevImage;
+            for (const objectUrl of objectUrls) {
+                const image = new Image();
+                image.src = objectUrl;
```
The `objectUrl` should be revoked after the image loaded.
If you await the images to be loaded, they can also all be inserted at the same time without flickering.
Might also avoid flickering more reliably if the first image is not the first to be loaded (in case this is currently an issue at all).

sisou replied:
They are inserted front to rear currently, so any flickering is not noticable.

sisou replied:
To expand on this: the images are inserted front-to-rear, in exactly the same place, so the front image hides all the other images behind it. Then they are fanned out in an animation. By the time the animation starts, all images are loaded and no flickering can be seen anymore.

sisou replied:
I added revoking of the objectUrls at the end, though.

danimoh commented on src/request/create/Create.js:
```diff
@@ -66,47 +59,47 @@ class Create {
         this._passwordSetter = new PasswordSetterBox($setPassword, { buttonI18nTag: 'passwordbox-confirm-create' });
-        this._downloadLoginFile = new DownloadLoginFile(
-            $downloadLoginFile,
-            I18n.translatePhrase('create-loginfile-any-device'),
```
The description in `DownloadLoginFile` is now not used anywhere anymore and can be removed from `DownloadLoginFile`.

danimoh commented on src/request/create/Create.js:
```diff
@@ -66,47 +59,47 @@ class Create {
+        this._selectedEntropy = Nimiq.Entropy.generate();
+        const masterKey = this._selectedEntropy.toExtendedPrivateKey();
+        const extPrivKey = masterKey.derivePath(request.defaultKeyPath);
+        this._selectedAddress = extPrivKey.toAddress().toUserFriendlyAddress();
```
There is no selection happening anymore, therefore the `_selected` can be stripped from names.

sisou replied:
Julian mentioned that we might bring selection back in the future.

danimoh commented on src/request/create/Create.js:
```diff
@@ -192,21 +176,23 @@ class Create {
 Create.Pages = {
-    CHOOSE_IDENTICON: 'choose-identicon',
-    CONFIRM_IDENTICON: 'confirm-identicon',
+    LOGIN_FILE_SELECTION: 'login-file-selection',
```
Not really a selection happening here.

danimoh commented on src/request/derive-polygon-address/DerivePolygonAddressApi.js:
```diff
@@ -0,0 +1,32 @@
+            throw new Errors.InvalidRequestError('Cannot derive a Polygon address for single-address accounts');
+        }
+        parsedRequest.keyLabel = this.parseLabel(request.keyLabel);
+        parsedRequest.polygonAccountPath = this.parsePath(request.polygonAccountPath, 'polygonAccountPath');
```
Should use `parsePolygonPath`.

danimoh commented on src/request/import/ImportApi.js:
```diff
@@ -24,6 +24,7 @@ class ImportApi extends BitcoinRequestParserMixin(TopLevelApi) {
             parsedRequest.expectedKeyId = (await this.parseKeyId(request.expectedKeyId)).id;
         }
         parsedRequest.bitcoinXPubPath = this.parseBitcoinPath(request.bitcoinXPubPath, 'bitcoinXPubPath');
+        parsedRequest.polygonAccountPath = this.parsePath(request.polygonAccountPath, 'polygonAccountPath');
```
Should use `parsePolygonPath`.

danimoh commented on src/request/sign-polygon-transaction/index.html:
```diff
@@ -0,0 +1,133 @@
+    <script defer bundle-polygon src="../../lib/polygon/PolygonContractABIs.js"></script>
+    <script defer bundle-polygon src="../../lib/polygon/PolygonRequestParserMixin.js"></script>
+
+    <script defer src="../../lib/IqonHash.js"></script>
+    <script defer src="../../lib/LoginFileConfig.js"></script>
+    <script defer src="../../lib/CurrencyInfo.js"></script>
```
Unused imports.

danimoh commented on src/request/sign-polygon-transaction/index.html:
```diff
@@ -0,0 +1,133 @@
+                </a> -->
+                <!-- <h1 data-i18n="sign-tx-heading-checkout" class="nq-h1 hide-standard hide-cashlink">Verify Payment</h1> -->
+                <h1 data-i18n="sign-tx-heading-tx" class="nq-h1 hide-checkout hide-cashlink">Confirm Transaction</h1>
+                <!-- <h1 data-i18n="sign-tx-heading-cashlink" class="nq-h1 hide-checkout hide-standard ">Create a Cashlink</h1> -->
```
Commented code and the hide classes can be removed.

danimoh commented on src/request/sign-polygon-transaction/SignPolygonTransactionApi.js:
```diff
@@ -0,0 +1,117 @@
+            request,
+            ['transfer', 'transferWithApproval', 'refund'],
+        );
+        parsedRequest.request = request.request;
```
More elegant than having the validation of `request.request` as a side effect of `parseOpenGsnForwardRequest` and assigning it afterwards, would be if `parseOpenGsnForwardRequest` would return both, the `description` and the `request`.

danimoh commented on client/src/PublicRequest.ts:
```diff
@@ -194,6 +207,32 @@ export type SignBtcTransactionRequest
+    /**
+     * For refund transactions from HTLCs the amount is not part of the relay request.
+     */
+    amount?: number,
```
For Nimiq and Bitcoin, the transaction amount is called `value`.

sisou replied:
I named it `amount`, because that's what the field is called in the contract method arguments for the other USDC methods.

danimoh commented on client/src/PublicRequest.ts:
```diff
@@ -194,6 +207,32 @@ export type SignBtcTransactionRequest
+    relayData: RelayData,
+
+    /**
+     * For refund transactions from HTLCs the amount is not part of the relay request.
```
Also for redeem transactions it's required.

And maybe rephrase to "forward request" as the type is also called like that.

danimoh commented on src/request/sign-polygon-transaction/SignPolygonTransactionApi.js:
```diff
@@ -0,0 +1,117 @@
+            if (!request.amount) {
+                throw new Errors.InvalidRequestError('`amount` required for refund method');
+            }
+        }
```
```suggestion
        // Check that amount exists when method is 'refund', and unset for other methods.
        if ((description.name === 'refund') !== !!request.amount) {
            throw new Errors.InvalidRequestError('`amount` required for refund method and ignored for other methods');
        }
```
`amount` is required only for refunds and ignored otherwise.

danimoh commented on src/request/sign-polygon-transaction/SignPolygonTransactionApi.js:
```diff
@@ -0,0 +1,117 @@
+            if (!request.approval) {
+                throw new Errors.InvalidRequestError('`approval` object required for transferWithApproval method');
+            }
+        }
```
```suggestion
        // Check that approval object exists when method is 'transferWithApproval', and unset for other methods.
        if ((description.name === 'transferWithApproval') !== !!request.approval) {
            throw new Errors.InvalidRequestError('`approval` object required for transferWithApproval method and ignored for other methods');
        }
```
`approval` is required only for `transferWithApproval` and ignored otherwise.

danimoh commented on client/src/PublicRequest.ts:
```diff
@@ -194,6 +207,32 @@ export type SignBtcTransactionRequest
+export type PolygonTransactionInfo = {
+    keyPath: string,
+
+    request: ForwardRequest,
```
Maybe better to call this `forwardRequest` to avoid having `request.request` in many places in the code.

danimoh commented on src/request/sign-polygon-transaction/SignPolygonTransaction.js:
```diff
@@ -0,0 +1,235 @@
+        const feeUnits = request.description.args.fee.toNumber();
+        if (feeUnits > 0) {
+            // For the fee, we do not display more than two decimals, as it would not add any value for the user
+            $fee.textContent = NumberFormatting.formatNumber(feeUnits / 1e6, 2, 2);
```
Can use `PolygonUtils` for conversions.

danimoh commented on src/request/sign-polygon-transaction/SignPolygonTransaction.js:
```diff
@@ -0,0 +1,235 @@
+            $fee.textContent = NumberFormatting.formatNumber(feeUnits / 1e6, 2, 2);
+            /** @type {HTMLDivElement} */
+            const $feeSection = (this.$el.querySelector('.fee-section'));
+            $feeSection.classList.remove('display-none');
```
Small fees that are rounded to `0.00` should be displayed as `< 0.01`.

sisou replied:
Only relevant in testnet, never in mainnet.

danimoh commented on src/request/sign-polygon-transaction/SignPolygonTransaction.js:
```diff
@@ -0,0 +1,235 @@
+     * @returns {Promise<void>}
+     * @private
+     */
+    async _onConfirm(request, resolve, reject, password) {
```
`request` doesn't have to be passed as it is available as instance property.
Alternatively, don't set `_request` on the instance and only use the passed `request` which slightly shortens the code.

mraveux committed:
Remove unnecessary new line

danimoh commented on src/request/sign-polygon-transaction/SignPolygonTransaction.css:
```diff
@@ -0,0 +1,90 @@
+#confirm-transaction .payment-info-line + .nq-card-header {
+    padding-top: 2rem;
+}
+
```
Unused.

danimoh commented on src/request/sign-polygon-transaction/SignPolygonTransaction.css:
```diff
@@ -0,0 +1,90 @@
+    text-align: center;
+    max-width: 100%;
+    overflow-wrap: break-word;
+}
```
Unused.

mraveux committed:
tsconfig: remove duplicate entry

mraveux commented on src/components/PolygonAddressInfo.css:
```diff
@@ -0,0 +1,114 @@
+    align-items: center;
+    opacity: 0.3;
+    transition: opacity 0.3s var(--nimiq-ease);
+}
```
I generally think more variables could be used in css to simplify code, especially when not using scss.
For example here `.short-address`, `.short-address span` and `.short-address .dots` have the same transition time since they **should** transition together. It then makes sense to link its transition-duration through a variable here, like
```suggestion
.polygon-address-info .short-address {
    --transition-duration: 0.3s;
    
    font-family: "Fira Mono", monospace;
    font-size: 2rem;
    line-height: 1.25;
    margin-top: 0.5rem;
    display: flex;
    align-items: center;
    white-space: nowrap;
    padding: 0.5rem 0.75rem;
    border-radius: 0.5rem;
    transition: background var(--transition-duration) var(--nimiq-ease);
    cursor: default;
}

.polygon-address-info .short-address span {
    opacity: 0.5;
    transition: opacity var(--transition-duration) var(--nimiq-ease);
}

.polygon-address-info .short-address .dots {
    display: flex;
    align-items: center;
    opacity: 0.3;
    transition: opacity var(--transition-duration) var(--nimiq-ease);
}
```
which would also make it easier to adapt the transition duration through the devtools, for finetuning or for designers to try themselves and give feedbacks.

And that works for any values that should be the same, or at least linked, through the dom tree (width = height, paddings, durations, colors, etc...). (since css variables only propagate down the tree, and not up)

sisou replied:
I applied the change suggested here in this comment. Please let me know any other occurrences where this can be done.

danimoh commented on src/request/sign-swap/index.html:
```diff
@@ -110,14 +116,15 @@ <h1 data-i18n="sign-swap-heading" class="nq-h1">Confirm Swap</h1>
                     <div class="account right-account">
                         <div class="identicon"></div>
                         <span class="label"></span>
+                        <span class="amount"></span>
```
This seems to be unused.
Was it meant to have class `new-balance` instead?
But I also didn't see the amount/balance being added in the design, so I believe it is superfluous.

mraveux commented on src/request/sign-polygon-transaction/index.html:
```diff
@@ -0,0 +1,133 @@
+            <svg height="48" width="54" color="inherit" class="loading-spinner"><g>
+                <path class="big-hex" d="M51.9,21.9L41.3,3.6c-0.8-1.3-2.2-2.1-3.7-2.1H16.4c-1.5,0-2.9,0.8-3.7,2.1L2.1,21.9c-0.8,1.3-0.8,2.9,0,4.2 l10.6,18.3c0.8,1.3,2.2,2.1,3.7,2.1h21.3c1.5,0,2.9-0.8,3.7-2.1l10.6-18.3C52.7,24.8,52.7,23.2,51.9,21.9z" stroke="currentColor" stroke-width="3" fill="none" stroke-linecap="round" opacity="0.4" stroke-dasharray="92.5 60"/>
+                <path class="small-hex" d="M51.9,21.9L41.3,3.6c-0.8-1.3-2.2-2.1-3.7-2.1H16.4c-1.5,0-2.9,0.8-3.7,2.1L2.1,21.9c-0.8,1.3-0.8,2.9,0,4.2 l10.6,18.3c0.8,1.3,2.2,2.1,3.7,2.1h21.3c1.5,0,2.9-0.8,3.7-2.1l10.6-18.3C52.7,24.8,52.7,23.2,51.9,21.9z" stroke="currentColor" stroke-width="3" fill="none" stroke-linecap="round" stroke-dasharray="47.5 105"/>
+            </g></svg>
```
The same attributes on a `<path/>` can be applied directly on the parent group `<g></g>` to prevent repetition
```suggestion
            <svg height="48" width="54" color="inherit" class="loading-spinner"><g stroke-width="3" fill="none" stroke-linecap="round" stroke="currentColor">
                <path class="big-hex" d="M51.9,21.9L41.3,3.6c-0.8-1.3-2.2-2.1-3.7-2.1H16.4c-1.5,0-2.9,0.8-3.7,2.1L2.1,21.9c-0.8,1.3-0.8,2.9,0,4.2 l10.6,18.3c0.8,1.3,2.2,2.1,3.7,2.1h21.3c1.5,0,2.9-0.8,3.7-2.1l10.6-18.3C52.7,24.8,52.7,23.2,51.9,21.9z" opacity="0.4" stroke-dasharray="92.5 60"/>
                <path class="small-hex" d="M51.9,21.9L41.3,3.6c-0.8-1.3-2.2-2.1-3.7-2.1H16.4c-1.5,0-2.9,0.8-3.7,2.1L2.1,21.9c-0.8,1.3-0.8,2.9,0,4.2 l10.6,18.3c0.8,1.3,2.2,2.1,3.7,2.1h21.3c1.5,0,2.9-0.8,3.7-2.1l10.6-18.3C52.7,24.8,52.7,23.2,51.9,21.9z" stroke-dasharray="47.5 105"/>
            </g></svg>
```

mraveux commented on src/request/derive-polygon-address/index.html:
```diff
@@ -0,0 +1,102 @@
+            <svg height="48" width="54" color="inherit" class="loading-spinner"><g>
+                <path class="big-hex" d="M51.9,21.9L41.3,3.6c-0.8-1.3-2.2-2.1-3.7-2.1H16.4c-1.5,0-2.9,0.8-3.7,2.1L2.1,21.9c-0.8,1.3-0.8,2.9,0,4.2 l10.6,18.3c0.8,1.3,2.2,2.1,3.7,2.1h21.3c1.5,0,2.9-0.8,3.7-2.1l10.6-18.3C52.7,24.8,52.7,23.2,51.9,21.9z" stroke="currentColor" stroke-width="3" fill="none" stroke-linecap="round" opacity="0.4" stroke-dasharray="92.5 60"/>
+                <path class="small-hex" d="M51.9,21.9L41.3,3.6c-0.8-1.3-2.2-2.1-3.7-2.1H16.4c-1.5,0-2.9,0.8-3.7,2.1L2.1,21.9c-0.8,1.3-0.8,2.9,0,4.2 l10.6,18.3c0.8,1.3,2.2,2.1,3.7,2.1h21.3c1.5,0,2.9-0.8,3.7-2.1l10.6-18.3C52.7,24.8,52.7,23.2,51.9,21.9z" stroke="currentColor" stroke-width="3" fill="none" stroke-linecap="round" stroke-dasharray="47.5 105"/>
+            </g></svg>
```
The same attributes on a `<path/>` can be applied directly on the parent group `<g></g>` to prevent repetition
```suggestion
            <svg height="48" width="54" color="inherit" class="loading-spinner"><g stroke-width="3" fill="none" stroke-linecap="round" stroke="currentColor">
                <path class="big-hex" d="M51.9,21.9L41.3,3.6c-0.8-1.3-2.2-2.1-3.7-2.1H16.4c-1.5,0-2.9,0.8-3.7,2.1L2.1,21.9c-0.8,1.3-0.8,2.9,0,4.2 l10.6,18.3c0.8,1.3,2.2,2.1,3.7,2.1h21.3c1.5,0,2.9-0.8,3.7-2.1l10.6-18.3C52.7,24.8,52.7,23.2,51.9,21.9z" opacity="0.4" stroke-dasharray="92.5 60"/>
                <path class="small-hex" d="M51.9,21.9L41.3,3.6c-0.8-1.3-2.2-2.1-3.7-2.1H16.4c-1.5,0-2.9,0.8-3.7,2.1L2.1,21.9c-0.8,1.3-0.8,2.9,0,4.2 l10.6,18.3c0.8,1.3,2.2,2.1,3.7,2.1h21.3c1.5,0,2.9-0.8,3.7-2.1l10.6-18.3C52.7,24.8,52.7,23.2,51.9,21.9z" stroke-dasharray="47.5 105"/>
            </g></svg>
```

danimoh commented on src/request/sign-swap/SignSwapApi.js:
```diff
@@ -68,6 +73,17 @@ class SignSwapApi extends BitcoinRequestParserMixin(TopLevelApi) {
+                description: /**
+                    @type {PolygonOpenDescription | PolygonOpenWithApprovalDescription}
+                */ (this.parseOpenGsnForwardRequest(request.fund, ['open', 'openWithApproval'])),
+                request: request.fund.request,
```
Similar to the comment for `SignPolygonTrasactionApi.parseOpenGsnForwardRequest` the method should return the parsed request and description at once, instead of parsing the request as a side effect.

mraveux commented on src/request/sign-polygon-transaction/SignPolygonTransaction.js:
```diff
@@ -0,0 +1,235 @@
+     */
+    constructor(request, resolve, reject) {
+        this._request = request;
+        /** @type {HTMLElement} */
+        this.$el = (document.getElementById(SignPolygonTransaction.Pages.CONFIRM_TRANSACTION));
```
In all of those situations if we are sure that the element exist we can cast the returned value of `getElementById` to be `HTMLElement` instead of `HTMLElement|null` which currently conflict with the type definition above it.
```suggestion
        /** @type {HTMLElement} */
        this.$el = /**@type {HTMLElement}*/(document.getElementById(SignPolygonTransaction.Pages.CONFIRM_TRANSACTION));
```
Or if we're not sure the element exist, we can also specify it by adding a `?` before the type, which also fix the type conflict.
```suggestion
        /** @type {?HTMLElement} */
        this.$el = document.getElementById(SignPolygonTransaction.Pages.CONFIRM_TRANSACTION));
```

sisou replied:
AFAIK this conflict is only in our IDEs, since it uses a newer TS version than the Keyguard code.
It compiles correctly.

This will be part of the changes required when updating the Keyguard's TS version.

danimoh commented on src/request/sign-swap/SignSwapApi.js:
```diff
@@ -42,8 +46,9 @@ class SignSwapApi extends BitcoinRequestParserMixin(TopLevelApi) {
                     recipientType: Nimiq.Account.Type.HTLC,
                     flags: Nimiq.Transaction.Flag.CONTRACT_CREATION,
                 }),
-                /** @type {string} */
-                senderLabel: (this.parseLabel(request.fund.senderLabel, false, 'fund.senderLabel')),
+                senderLabel: /** @type {string} */ (this.parseLabel(
+                    request.fund.senderLabel, false, 'fund.senderLabel',
+                )),
```
Why did you change the order of the type casts in this file? The other way is how we've typically done it and reads nicer.

sisou replied:
Probably because VSCode with its newer TS version complains that it's invalid.

The old style is not compatible with new TS versions anymore and will have to be changed everywhere when we upgrade.

mraveux reviewed (approved):
I pushed some smol commits with very minor changes, and added some comment on some small things I think can be improved not only in this PR but overall in the keyguard. Otherwise I didn't found anything breaking or not working as expected.

Good job 👍

danimoh commented on client/src/PublicRequest.ts:
```diff
@@ -236,6 +275,9 @@ export type SignSwapRequestCommon = SimpleRequest & {
         }>
+    ) | (
+        {type: 'USDC'}
+        & PolygonTransactionInfo
```
Omit `amount` because it's not required and ignored.

danimoh commented on src/request/sign-swap/SignSwapApi.js:
```diff
@@ -144,40 +173,50 @@ class SignSwapApi extends BitcoinRequestParserMixin(TopLevelApi) {
 
         if (request.layout === SignSwapApi.Layouts.SLIDER && parsedRequest.layout === SignSwapApi.Layouts.SLIDER) {
             // SLIDER layout is only allowed for NIM-BTC swaps
```
```suggestion
            // SLIDER layout is only allowed for crypto-to-crypto swaps
```

danimoh commented on src/request/sign-swap/SignSwapApi.js:
```diff
@@ -144,40 +173,50 @@ class SignSwapApi extends BitcoinRequestParserMixin(TopLevelApi) {
             };
+            parsedRequest.polygonAddresses = request.polygonAddresses.map((address, index) => ({
+                address: this.parsePolygonAddress(address.address, `polygonAddresses[${index}].address`),
+                balance: this.parsePositiveInteger(address.balance, true, `polygonAddresses[${index}].balance`),
```
```suggestion
            parsedRequest.polygonAddresses = request.polygonAddresses.map(({ address, balance }, index) => ({
                address: this.parsePolygonAddress(address, `polygonAddresses[${index}].address`),
                balance: this.parsePositiveInteger(balance, true, `polygonAddresses[${index}].balance`),
```

danimoh commented on src/request/sign-swap/SignSwapApi.js:
```diff
@@ -144,40 +173,50 @@ class SignSwapApi extends BitcoinRequestParserMixin(TopLevelApi) {
+                }
             }
+
+            // TODO Verify that used Polygonaddress is the one in polygonAddresses[]
```
:point_up_2: TODO

And ideally also check the polygon account's provided balance similar to how we do it above for Nimiq.

danimoh commented on client/src/PublicRequest.ts:
```diff
@@ -194,6 +207,32 @@ export type SignBtcTransactionRequest
     = SignBtcTransactionRequestStandard
     | SignBtcTransactionRequestCheckout;
 
+export type PolygonTransactionInfo = {
```
Might also want to add a prop `currency` to make clear via the request which currency should be moved.
Currently, that would always be `usdc` of course.

sisou replied:
Yeah, this is a bit of a conundrum. Keyguard only supports USDC, because it only supports sending via our custom transfer contract, which only supports USDC and we only show USDC as the currency ticker in the UI, cannot show anything else.

So while everything runs on Polygon, it can only reasonably interact with USDC. So should we rename all types from Polygon* to Usdc*?

sisou replied:
The `parseOpenGsnForwardRequest` methods in the sign-polygon-transaction and sign-swap API classes validate that the recipient of any transaction is a known USDC smart contract address. Any other recipient throws. So I think that covers it.

danimoh commented on src/request/sign-swap/SignSwapApi.js:
```diff
@@ -218,6 +257,80 @@ class SignSwapApi extends BitcoinRequestParserMixin(TopLevelApi) {
+    /**
+     *
+     * @param {KeyguardRequest.PolygonTransactionInfo} request
+     * @param {string[]} allowedMethods
```
```suggestion
     * @param {Array<'open' | 'openWithApproval' | 'redeem' | 'redeemWithSecretInData'>} allowedMethods
```
Only a subset of possible methods is supported by this method.

danimoh commented on src/request/sign-polygon-transaction/SignPolygonTransactionApi.js:
```diff
@@ -0,0 +1,117 @@
+    /**
+     *
+     * @param {KeyguardRequest.PolygonTransactionInfo} request
+     * @param {string[]} allowedMethods
```
```suggestion
     * @param {Array<'transfer' | 'transferWithApproval' | 'refund'>} allowedMethods
```
Only a subset of possible methods is supported by this method.

danimoh commented on types/Keyguard.d.ts:
```diff
@@ -55,6 +55,104 @@ type ParsedBitcoinTransactionInput = {
+    readonly sigR: string,
+    readonly sigS: string,
+    readonly sigV: ethers.BigNumber,
+}
```
Can extend `PolygonTransferArgs` to avoid code duplication.

danimoh commented on types/Keyguard.d.ts:
```diff
@@ -55,6 +55,104 @@ type ParsedBitcoinTransactionInput = {
+    readonly sigR: string,
+    readonly sigS: string,
+    readonly sigV: ethers.BigNumber,
+}
```
Can extend `PolygonOpenArgs` to avoid code duplication.

danimoh commented on src/request/sign-swap/SignSwapApi.js:
```diff
@@ -218,6 +257,80 @@ class SignSwapApi extends BitcoinRequestParserMixin(TopLevelApi) {
+            if (!request.approval) {
+                throw new Errors.InvalidRequestError('`approval` object required for openWithApproval method');
+            }
+        }
```
```suggestion
        // Check that approval object exists when method is 'openWithApproval', and unset for other methods.
        if ((description.name === 'openWithApproval') !== !!request.approval) {
            throw new Errors.InvalidRequestError('`approval` object required for openWithApproval method and ignored for other methods');
        }
```
`approval` is required only for `openWithApproval` and ignored otherwise.

danimoh commented on src/request/sign-swap/SignSwap.js:
```diff
@@ -56,26 +62,36 @@ class SignSwap {
+        /** @type {HTMLSpanElement} */
         const $rightLabel = (this.$el.querySelector('.right-account .label'));
         /** @type {HTMLDivElement} */
+        const $rightNewBalance = (this.$el.querySelector('.right-account .new-balance'));
```
```suggestion
        const $rightNewBalance = ($rightAccount.querySelector('.new-balance'));
```
And similarly for the others.

danimoh commented on src/request/sign-swap/SignSwap.js:
```diff
@@ -180,6 +212,17 @@ class SignSwap {
+                        // is the transaction value + tx fee.
+                        ? redeemTx.amount + redeemTx.description.args.fee.toNumber() + request.redeemFees.funding
+                        : 0; // Should never happen, if parsing works correctly
+                break;
```
Unnecessary code duplication. The entire second switch block can be avoided by putting it into a method which is called for `exchangeBaseAsset` and `exchangeOtherAsset`.

danimoh commented on src/request/sign-swap/SignSwap.js:
```diff
@@ -201,11 +244,22 @@ class SignSwap {
+            0,
+            this._assetDecimals(exchangeOtherAsset) - exchangeRateDigitsLength,
+        );
+        $exchangeRate.textContent = `1 ${exchangeBaseAsset} = ${NumberFormatting.formatNumber(
```
What was wrong with the old implementation?

sisou replied:
Keyguard always showed less decimals than the Wallet in the UI.

danimoh replied:
Is it correct though?
Wouldn't the `.replace('0', '')` line lead to an undesired/wrong result if exchangeRate is a non-zero number that includes a zero in it's integer part?
Was it `.replace(/^0/, '')` that you wanted to achieve?

sisou replied:
Yes, that's right. Fixing it, and also fixing it wherever I copy-pasta'd this code from.

danimoh commented on src/request/sign-swap/SignSwap.js:
```diff
@@ -318,31 +386,115 @@ class SignSwap {
+                    ? fundTx.transaction.sender.toUserFriendlyAddress()
+                    : redeemTx.type === 'NIM'
+                        ? redeemTx.transaction.recipient.toUserFriendlyAddress()
+                        : ''; // Should never happen, if parsing works correctly
```
This is the same as `swapNimAddress` which was already calculated above.

danimoh commented on src/request/sign-swap/SignSwap.js:
```diff
@@ -318,31 +386,115 @@ class SignSwap {
+                    newBalance: address.address === activeAddress
+                        ? newBalance
+                        : address.balance,
+                }));
```
```suggestion
                const segments = request.nimiqAddresses.map({ address, balance } => ({
                    address: address,
                    balance:  balance,
                    active: address === activeAddress,
                    newBalance: address === activeAddress ? newBalance : balance,
                }));
```

danimoh commented on src/request/sign-swap/SignSwap.js:
```diff
@@ -451,14 +606,92 @@ class SignSwap {
+                    /* bytes32 sigR */ sigR,
+                    /* bytes32 sigS */ sigS,
+                    /* uint8 sigV */ sigV,
+                ]);
```
Very similar to approval signing in `SignPolygonTransaction`. Can probably be refactored into a common method living in `PolygonUtils`.

danimoh commented on src/request/sign-polygon-transaction/SignPolygonTransaction.js:
```diff
@@ -0,0 +1,235 @@
+
+            const functionSignature = usdcContract.interface.encodeFunctionData(
+                'approve',
+                [CONFIG.USDC_TRANSFER_CONTRACT_ADDRESS, this._request.description.args.approval],
```
We currently approve the maximum possible amount, correct? I'm a bit uncomfortable with this.
It's fine while the feature is behind the trial password, but I think for the real release it should be limited to the transferred amount and the Keyguard should check that.
At least until the transfer contract and htlc contract are audited.

danimoh commented on src/request/sign-swap/SignSwap.css:
```diff
@@ -25,31 +26,26 @@
+    margin: 2rem auto 0 auto !important;
+    column-gap: 1rem;
+    font-weight: 600;
+    height: 21px;
```
Should be using rem.

danimoh commented on src/request/swap-iframe/SwapIFrameApi.js:
```diff
@@ -215,6 +224,51 @@ class SwapIFrameApi extends BitcoinRequestParserMixin(RequestParser) { // eslint
+                value: 0,
+            }));
+
+            if (description.name !== 'open') {
```
Is the htlc data we get always for `open` and never for `openWithApproval`?
Might want to add a comment for that and that the pre-signed approval is attached below.

sisou replied:
yes.

danimoh commented on src/request/swap-iframe/SwapIFrameApi.js:
```diff
@@ -215,6 +224,51 @@ class SwapIFrameApi extends BitcoinRequestParserMixin(RequestParser) { // eslint
+        if (request.redeem.type === 'USDC' && storedRequest.redeem.type === 'USDC') {
+            redeem = {
+                type: 'USDC',
+                htlcId: `0x${Nimiq.BufferUtils.toHex(Nimiq.BufferUtils.fromAny(request.redeem.htlcId.substring(2)))}`,
```
Add a comment of what is being cut off here. If your intention is to cut off a leading `0x`, `.replace(/^0x/i, '')` would be more reliable as it would verify an invalid prefix and still allow base 64 encoding of the input `htlcId`.

danimoh commented on src/request/swap-iframe/SwapIFrameApi.js:
```diff
@@ -215,6 +224,51 @@ class SwapIFrameApi extends BitcoinRequestParserMixin(RequestParser) { // eslint
+                type: 'USDC',
+                htlcId: `0x${Nimiq.BufferUtils.toHex(Nimiq.BufferUtils.fromAny(request.redeem.htlcId.substring(2)))}`,
+                htlcDetails: {
+                    hash: Nimiq.BufferUtils.toHex(Nimiq.BufferUtils.fromAny(request.redeem.hash)),
```
I guess here a leading `0x` should also be removed if present?

sisou replied:
`hash` has no leading 0x at this place anymore, it's cut off before. only the `htlcId` has it.

danimoh commented on src/request/swap-iframe/SwapIFrameApi.js:
```diff
@@ -215,6 +224,51 @@ class SwapIFrameApi extends BitcoinRequestParserMixin(RequestParser) { // eslint
+                htlcId: `0x${Nimiq.BufferUtils.toHex(Nimiq.BufferUtils.fromAny(request.redeem.htlcId.substring(2)))}`,
+                htlcDetails: {
+                    hash: Nimiq.BufferUtils.toHex(Nimiq.BufferUtils.fromAny(request.redeem.hash)),
+                    timeoutTimestamp: this.parsePositiveInteger(request.redeem.timeout, false, 'redeem.timeout'),
```
Is this an actual timestamp or a block number? What's the unit? Is it the same as for EUR (because they're compared below)?

Checking that it's not in the past according to the user's time could also make sense.

sisou replied:
Yes, unix timestamp in seconds. That's why the property is called `timeoutTimestamp` :wink:

danimoh commented on src/request/swap-iframe/SwapIFrameApi.js:
```diff
@@ -241,22 +295,27 @@ class SwapIFrameApi extends BitcoinRequestParserMixin(RequestParser) { // eslint
+            ? 'timeoutTimestamp' in fund.htlcDetails
+                ? fund.htlcDetails.timeoutTimestamp
+                : undefined
+            : fund.description.args.timeout.toNumber();
```
Same here, is this an actual timestamp or a block number?

sisou replied:
HTLC contract timeouts are timestamps.

danimoh commented on src/request/swap-iframe/SwapIFrameApi.js:
```diff
@@ -473,6 +539,55 @@ class SwapIFrameApi extends BitcoinRequestParserMixin(RequestParser) { // eslint
+                typedData.message,
+            );
+
+            result.usdc = {
+                message: typedData.message,
+                signature,
+            };
```
Why are we not pre-signing the refund transaction here?

sisou replied:
Because it has a high likelihood that it will be invalid (i.e. the relay doesn't work anymore or doesn't have enough balance), so I did not implement it (yet). 

We found that in mainnet the relays are more reliable, so makes sense to add it now.

sisou replied:
Thinking about it, and with added experience from mainnet, it's the _network_ tx fee that is fluctuating a lot between blocks and between minutes. We already increased the fee (percentage-wise) for redeem transactions to avoid them failing too often. The refund transaction is only valid even later (>2h), so setting the fee for it already at swap-start is pure guess-work, and has a high likelihood that the tx fee is too low, or the user overpays.

danimoh commented on src/request/sign-swap/SignSwap.js:
```diff
@@ -451,14 +606,92 @@ class SignSwap {
+                    /* bytes32 sigR */ sigR,
+                    /* bytes32 sigS */ sigS,
+                    /* uint8 sigV */ sigV,
+                ]);
```
Could consider moving this to `SwapIframeApi` where all the other signing is happening.

sisou replied:
But it _can_ be signed already without the confirmed swap. Everything in SwapIframeApi can only be signed with details gotten from a confirmed swap, so it makes sense to sign this here and the rest there.

danimoh commented on src/request/swap-iframe/SwapIFrameApi.js:
```diff
@@ -473,6 +539,55 @@ class SwapIFrameApi extends BitcoinRequestParserMixin(RequestParser) { // eslint
+
+            const wallet = new ethers.Wallet(privateKeys.usdc);
+
+            const signature = await wallet._signTypedData(
+                typedData.domain,
+                /** @type {Record<string, ethers.ethers.TypedDataField[]>} */ (/** @type {unknown} */ (cleanedTypes)),
+                typedData.message,
+            );
```
Could verify the signature via `ethers.utils.verifyTypedData`.

sisou replied:
I don't see any use in verifying a signature with the same library that created it, with the exact same inputs. I trust the library that it has tests for that. Experience in production over the last months also shows us that the signatures are valid. We are also not verifying the NIM and BTC signatures.

danimoh commented on src/request/swap-iframe/SwapIFrameApi.js:
```diff
@@ -573,6 +688,46 @@ class SwapIFrameApi extends BitcoinRequestParserMixin(RequestParser) { // eslint
+                typedData.domain,
+                /** @type {Record<string, ethers.ethers.TypedDataField[]>} */ (/** @type {unknown} */ (cleanedTypes)),
+                typedData.message,
+            );
```
`parsedRequest.redeem.htlcDetails` are not part of the signed data and the signed data is not cross checked with the `htlcDetails`, and therefore unrelated `htlcDetails` can be passed to the Keyguard without the Keyguard noticing ~~which makes passing and verifying them somewhat unnecessary~~ (edit: not actually unnecessary: while the check is not too insightful within the Keyguard, the Hub can rely on the fact that the Keyguard cross checked the Hub's data).

There might also be similar issues for the amount and potentially other things.

I guess to really check them, the htlc would need to be fetched by its id. Which I'm not sure whether we'd want to do that or whether it's necessary even. If the user's funding transaction is only sent out in the Hub after the other side set up the HTLC and checking that HTLC against the data passed to the Keyguard (and which the Keyguard verified), it's probably fine.

To what extend is the hub checking the to be redeemed usdc htlc created by fastspot, before sending out our signed funding tx of the other currency?

Edit: after looking at libswap, I think the created htlc is not checked sufficiently. I believe we'd need to check the `recipient`, `hash` and `timeout` here: https://github.com/nimiq/libswap-js/blob/271705871d56b333fa0b01f2e81ed1febc8c2692/src/UsdcAssetAdapter.ts#L139

sisou replied:
Everything is checked in the Hub's `SetupSwapSuccess`: the correct recipient address, the hash(Roots) and also the timeouts.

danimoh replied:
I don't think that this solves the problem described above, and is also not entirely true.

The Hub receives the USDC HTLC info from Fastspot and forwards it to the Keyguard.
- Both, the Hub and Keyguard check whether the hash specified in that info matches the other swap contract.
- The Hub additionally checks whether the HTLC info has the expected redeem address or refund address. The Keyguard doesn't but should.
- The Keyguard to some degree checks the timeouts. The Hub doesn't but should.

But all that doesn't really matter because
- neither the Hub nor the Keyguard check the HTLC info against the actual HTLC contract data `htlcData`.
- the HTLC data is not contained in the data the Keyguard actually signs, as only the `htlcId` is contained in the signature.

I.e. the HTLC details, the HTLC contract data and the HTLC id can be completely unrelated and Fastspot could arbitrarily lie to us about what info we're signing and the details it claims we're signing, which neither the Hub nor the Keyguard check.

I don't know how the HTLC id is computed and whether we have all the data required to compute it. If we do, obviously the correct thing to do would be to recalculate it and check whether it matches. If we can't verify it, then the only thing we can do is checking whether the contract that Fastspot set-up on the chain actually contains what we expect (hash, recipient, timeout, etc.) during execution of the swap in the Wallet, before we send out our funding transaction, which comes back to the checks in libswap I already mentioned in the opening post.

It is actually a general problem that we do not or can not verify the HTLC id or address.
- For Nimiq, we can not verify the HTLC address, because the computation of the contract address includes the Fastspot sender address and validity start height which we don't now. However, libswap checks whether Fastspot's incoming funding transaction includes the correct htlc contract data, which was previously checked by the Keyguard. By this, the correctness of the HTLC address for the HTLC data is verified.
- For Bitcoin, the Keyguard only receives the UTXO of the HTLC funding transaction (transaction hash and output index), which the Hub fetched from the chain for the HTLC address, but which the Keyguard can not verify without fetching the chain itself. The HTLC address should be possible to be computed from the HTLC script to check that the HTLC address we got from Fastspot is correct, but the Hub currently doesn't do so. Neither does libswap, thus for Bitcoin, the correctness of the HTLC address for the HTLC data is _not_ verified.
- Fot USDC, as explained above, the correctness of the HTLC id is not verified in any way in the Keyguard, Hub or libswap. We would need to verify it in the Keyguard or Hub by recalculating it if possible, or otherwise in libswap by checking for Fastspot's incoming funding transaction to include the correct redeem address, hash and timeout.

Not verifying the correctness undermines the integrity of the atomic swap and would allow Fastspot to steal the user's funds by sending him the proper HTLC info the Hub and Keyguard would expect and verify, but an unrelated HTLC address or id for a HTLC the user can't claim. If the user then funds his part of the HTLC, Fastspot can claim his funds without the user being able to claim the redeem transaction. After the timeout, Fastspot would then be able to claim its own HTLC back, effectively stealing the user's funds.

danimoh commented on src/request/swap-iframe/SwapIFrameApi.js:
```diff
@@ -573,6 +688,46 @@ class SwapIFrameApi extends BitcoinRequestParserMixin(RequestParser) { // eslint
+
+            const wallet = new ethers.Wallet(privateKeys.usdc);
+
+            const signature = await wallet._signTypedData(
+                typedData.domain,
+                /** @type {Record<string, ethers.ethers.TypedDataField[]>} */ (/** @type {unknown} */ (cleanedTypes)),
+                typedData.message,
+            );
```
Could verify the signature via `ethers.utils.verifyTypedData`.

sisou replied:
Same reasoning as [above](https://github.com/nimiq/keyguard/pull/465#discussion_r1358902067).

danimoh commented on src/translations/es.json:
```diff
@@ -14,6 +14,7 @@
     "back-to-cpl": "Volver a CryptoPayment.link",
 
     "funding-cashlink": "Fundando cashlink",
+    "label-unknown": "Deconocido",
```
```suggestion
    "label-unknown": "Desconocido",
```

sisou replied:
Why is this in the review? This word comes from Transifex :laughing:

danimoh commented on tests/lib/PolygonKey.spec.js:
```diff
@@ -0,0 +1,99 @@
+/* global PolygonConstants */
+/* global Dummy */
+
+describe('BitcoinKey', () => {
```
```suggestion
describe('PolygonKey', () => {
```

danimoh commented on tests/lib/PolygonKey.spec.js:
```diff
@@ -0,0 +1,99 @@
+        //     index: 25, // Our output was the 26th in the input tx
+        //     witnessUtxo: {
+        //         script: BitcoinJS.Buffer.from('001484eb9bcbd90ce7d3360992259e4b9b818215a960', 'hex'),
+        //         value: 27738766,
+        //     },
+        // }].sort((input1, input2) => {
+        //     if (input1.hash !== input2.hash) return input1.hash < input2.hash ? -1 : 1;
+        //     return input1.index - input2.index;
+        // });
+
+        // const sum = inputs.reduce((sum, input) => sum + input.witnessUtxo.value, 0);
+        // const amount = Math.round(sum / 2);
+
+        // const paths = [`${BitcoinConstants.ACCOUNT_DERIVATION_PATH.BIP84.TEST}/0/0`];
+
+        // const outputs = [{
+        //     address: 'tb1qm37aps59r86g79jg5xh8ecgfqfncv92wzfyyx7',
+        //     value: amount,
+        // }, {
+        //     address: 'tb1qxvrezy277f2035kasl39pls454yjwtgqkfv05z',
+        //     value: sum - amount - 481, // 481 is the fee
+        // }].sort((output1, output2) => {
+        //     return (output1.value - output2.value) || (output1.address < output2.address ? -1 : 1);
+        // });
+
+        // const psbt = new BitcoinJS.Psbt({ network: BitcoinJS.networks.testnet });
+
+        // // @ts-ignore Argument of type 'import("...").Buffer' is not assignable to parameter of type 'Buffer'.
+        // psbt.addInputs(inputs);
+        // psbt.addOutputs(outputs);
+
+        // btcKey.sign(paths, psbt);
+
+        // psbt.finalizeAllInputs();
+
+        // const tx = psbt.extractTransaction();
+
+        // expect(tx.toHex()).toEqual(
+        //     '0200000000010632eb675c33a5bcecc2569d80046aa7d01eb32e502a54b611cd4c619e63f583160100000000ffffffff5d0b94b92fdfdbfa756f17bc70e6f42fa2e7fbc535c92b4113966cb002a6d7180000000000ffffffff82c051201085b0b7f642702a138bc0628c09a312215f5fc4e1ac03e313ce0b550100000000ffffffffdf2fce5028c75de70599f99122da561e41b75b39761f358ee1999722f5aeed5f0000000000ffffffff4707888e79444b41d07440fc0b94c330b8ef7d41d ...
+        // );
```
:smile:

danimoh commented on tools/build.sh:
```diff
@@ -129,6 +128,17 @@ if [ ! $? -eq 0 ]; then
+echo "$ethersjs_hashsum" | ${SHA256SUM} --check
+
+if [ ! $? -eq 0 ]; then
+    output "💥  BitcoinJS file integrity check failed!"
```
```suggestion
    output "💥  EthersJS file integrity check failed!"
```

danimoh commented on types/Keyguard.d.ts:
```diff
@@ -55,6 +55,104 @@ type ParsedBitcoinTransactionInput = {
+    readonly approval: ethers.BigNumber,
+    readonly sigR: string,
+    readonly sigS: string,
+    readonly sigV: ethers.BigNumber,
```
These could also be moved into a common base type.

And might want to add a comment that `approval` refers to the approved amount.

danimoh commented on types/Keyguard.d.ts:
```diff
@@ -99,7 +197,11 @@ type ConstructSwap<T extends KeyguardRequest.SignSwapRequestCommon> = Transform<
             refundAddress: string,
-        } | {
+        } | (
+            { type: 'USDC' } & KeyguardRequest.PolygonTransactionInfo & {
```
```suggestion
            { type: 'USDC' } & Omit<KeyguardRequest.PolygonTransactionInfo, 'amount'> & {
```
I believe the amount can be omitted.

Can also use the `Transform` utility to make this a bit nicer.

danimoh commented on types/Keyguard.d.ts:
```diff
@@ -121,7 +223,12 @@ type ConstructSwap<T extends KeyguardRequest.SignSwapRequestCommon> = Transform<
             output: KeyguardRequest.BitcoinTransactionChangeOutput,
-        } | {
+        } | (
+            { type: 'USDC' } & Omit<KeyguardRequest.PolygonTransactionInfo, 'approval'> & {
```
```suggestion
            { type: 'USDC' } & Omit<KeyguardRequest.PolygonTransactionInfo, 'approval' | 'amount'> & {
```
Should the optional `amount` be omitted as it gets added as required property, or does it make no difference in the resulting type?

danimoh commented on types/Keyguard.d.ts:
```diff
@@ -197,6 +310,10 @@ type Parsed<T extends KeyguardRequest.Request> =
             },
+            polygonAddresses: Array<{
+                address: string,
+                balance: number, // cents
```
Not cents.

sisou replied:
Then what?

danimoh committed:
Usdc code cleanup

See PR#465

danimoh committed:
DerivePolygonAddresApi,ImportApi: use parsePolygonPath for parsing polygonAccountPath

danimoh committed:
PolygonRequestParserMixin: parse forward request data as hex

danimoh committed:
SignPolygonTransactionApi,SignSwapApi: allow and require amount and approval only iff they are needed

danimoh committed:
NumberFormatting: simplify formatNumber and avoid trailing decimal separator

danimoh committed:
internationalization: update Chinese translations

danimoh commented on src/request/derive-polygon-address/DerivePolygonAddress.js:
```diff
@@ -0,0 +1,90 @@
+        TopLevelApi.setLoading(true);
+        const passwordBuffer = password && password.length > 0
+            ? Utf8Tools.stringToUtf8ByteArray(password)
+            : undefined;
```
```suggestion
        const passwordBuffer = password ? Utf8Tools.stringToUtf8ByteArray(password) : undefined;
```
Password length check already covered by just checking for `password`.

danimoh reviewed (changes requested):
Lots of fancy stuff, nice.

I'm leaving a couple of review comments and marking this as "requesting changes" which shouldn't stop it from being merged though.
I already fixed many of the straight forwards things on the `usdc-review` branch, and marked those things with the eyes emoji here :eyes: 
I did not yet test those changes yet though. Feel free to review those fixes too and fast-forward merge them into this branch. Otherwise, I'll do it later when I get to testing them. Otherwise, it can also still be merged after the release of the password protected version.

There are two or three things still missing in my opinion, which should happen at some point when it's live on mainnet. It's not necessarily required for the password protected version though.

- ethers and OpenGSN are *a lot* of external and unreviewed code being added to the Keyguard. I'm still an advocate for putting them into a sandboxed iframe.
- I'm unconfortable with the user giving approvals for the maximum possible amount. I think until the smart contracts are more tested and audited, we should only approve the required amount and have the Keyguard also check that only the required amount is being approved. I asked Jeff to audit the smart contracts. Once that is done I think we can consider approving the maximum again, but still not a huge fan of this, but also being aware of the higher incurring fees for separate approvals for each transaction of course.
- The htlc redeeming is not much verified. Not sure yet whether that's any issue at all. Currently a htlc hash and timestamp unrelated to the actual htlc being signed can be passed without the Keyguard noticing. But if the user's funding transaction is only sent out *after* the other side set up the HTLC and checking that HTLC against the data passed to the Keyguard (and which the Keyguard verified), it's probably fine. ~~I didn't look yet at whether the Hub verifies these things (but probably it does).~~ (Edit: after looking at libswap, I think the created htlc is not checked sufficiently. I believe we'd need to check the `recipient`, `hash` and `timeout` here: https://github.com/nimiq/libswap-js/blob/271705871d56b333fa0b01f2e81ed1febc8c2692/src/UsdcAssetAdapter.ts#L139). Additionally, adding a bit of clarifying commentary to the code about the involved thoughts and concerns would probably be a good idea.

danimoh reviewed (changes requested):
[Duplicate comment which was posted on page reload after github was erroring with a message that the page takes too long to load when submitting the comments :D]

danimoh committed:
client: update typescript

Update typescript such that projects can add the Keyguard client via gitpkg without it failing
to build if they include other dependencies which use newer typescript syntax, which somehow
affect the client build.

Also setting the tsconfig.json target to es2021 as our rollup version is not compatible with
the current esnext yet.

sisou committed:
Improve error messages for disallowed fields

sisou committed:
Re-add missing scripts to sign-polygon-transaction

sisou committed:
Move swap exchange rate into SwapFeesTooltip

sisou committed:
Avoid naming USDC smallest unit "cents"

sisou committed:
Use a CSS variable for related transition durations

sisou committed:
Reduce repetition in loading spinner SVG

sisou committed:
Rename USDC address info `balance'  field to `usdcBalance`

To better reflect that this balance relates to the USDC token, not Polygon's native currency.

sisou committed:
Rename OpenGSN interface imports

sisou committed:
Minify OpenGSN library

sisou committed:
Extract common crypto utilities into their own library

sisou committed:
Revoke objectUrls in Create after images have loaded

sisou committed:
Change internal page name for first screen of the Create flow

To reflect that no LoginFile selection is happening anymore.

sisou committed:
Use parsed OpenGSN ForwardRequest from parser as return value

sisou committed:
For swaps, verify details of used Polygon address

sisou committed:
Deduplicate exchange rate values compute

sisou committed:
Deduplicate swap NIM address computation

sisou committed:
Add explaining comment, improve hex parser in SwapIFrameApi

sisou committed:
Extract USDC approval signing into method on PolygonKey

danimoh commented on src/request/sign-swap/SignSwapApi.js:
```diff
@@ -143,40 +181,71 @@ class SignSwapApi extends BitcoinRequestParserMixin(TopLevelApi) {
+            const polygonAddress = parsedRequest.fund.type === 'USDC'
+                ? parsedRequest.fund.request.from
+                : parsedRequest.redeem.type === 'USDC'
+                    ? parsedRequest.redeem.request.from
```
Little comment here would help, why it's `parsedRequest.redeem.request.from` and not `parsedRequest.redeem.request.to`.

danimoh commented on src/lib/polygon/PolygonKey.js:
```diff
@@ -0,0 +1,145 @@
+
+    /**
+     * @param {string} path
+     * @param {ethers.Contract} usdcContract
```
The USDC contract passed is the same in both cases where this method is used, and would probably also be the same in other potential use-cases. It would thus not need to be passed as a parameter, but could be created within the method.

sisou replied:
I did it this way because creating the contract in the `PolygonKey.js` file means adding a dependency on the `PolygonContractABIs.js` file, which would then be required for all places where a `PolygonKey` is instatiated, but not used to sign, like Create/Import/DerivePolygonAddress, where the additional bundled ABIs are only dead weight.

danimoh commented on src/lib/polygon/PolygonKey.js:
```diff
@@ -0,0 +1,145 @@
+            types,
+            message,
+        );
+
```
This little extra-check of the signature was removed with the refactoring:

```suggestion

        const signerAddress = ethers.utils.verifyTypedData(domain, types, message, signature);
        if (signerAddress !== fromAddress) {
            reject(new Errors.CoreError('Failed to sign approval'));
            return;
        }

```

[But seems that you don't like it](https://github.com/nimiq/keyguard/pull/465#discussion_r1358902067), which is then also fine.

sisou replied:
Ha, you spotted that :D 

Yes, I think it's unnecessary now that we know it works.

danimoh reviewed (approved):
Thanks for incorporating the changes.

At first, I thought the use of `parsedRequest.redeem.request.from` would be wrong (see comment below), but it seems to be fine?

I believe this now covers the previous review comments, and the PR is ready to be merged :tada:

sisou committed:
Add explaining comment on why for Polygon HTLC redemption the user's address is still the sender
